tools/afl/src/smessage_encrypt_roundtrip.c

```diff
@@ -0,0 +1,171 @@
+                                        &encrypted_size);
+
+    if (err != THEMIS_BUFFER_TOO_SMALL) {
+        fprintf(stderr, "failed to determine encrypted message size: %d\n", err);
+        return 2;
+    }
+
+    encrypted_bytes = malloc(encrypted_size);
+    if (!encrypted_bytes) {
+        fprintf(stderr, "failed to allocate memory for encrypted message (%zu bytes)\n", encrypted_size);
+        return 2;
+    }
+
+    err = themis_secure_message_encrypt(private_key_bytes,
+                                        private_key_size,
+                                        public_key_bytes,
+                                        public_key_size,
+                                        message_bytes,
+                                        message_size,
+                                        encrypted_bytes,
+                                        &encrypted_size);
+
+    if (err != THEMIS_SUCCESS) {
+        fprintf(stderr, "failed to encrypt message: %d\n", err);
+        return 2;
+    }
+
+    /*
+     * Then try decrypting it back.
+     */
+
+    uint8_t* decrypted_bytes = NULL;
+    size_t decrypted_size = 0;
+
+    err = themis_secure_message_decrypt(private_key_bytes,
+                                        private_key_size,
+                                        public_key_bytes,
+                                        public_key_size,
+                                        encrypted_bytes,
+                                        encrypted_size,
+                                        decrypted_bytes,
+                                        &decrypted_size);
+
+    if (err != THEMIS_BUFFER_TOO_SMALL) {
+        fprintf(stderr, "failed to determine decrypted message size: %d\n", err);
+        return 3;
+    }
+
+    decrypted_bytes = malloc(decrypted_size);
+    if (!decrypted_bytes) {
+        fprintf(stderr, "failed to allocate memory for decrypted message (%zu bytes)\n", decrypted_size);
+        return 3;
+    }
+
+    err = themis_secure_message_decrypt(private_key_bytes,
+                                        private_key_size,
+                                        public_key_bytes,
+                                        public_key_size,
+                                        encrypted_bytes,
+                                        encrypted_size,
+                                        decrypted_bytes,
+                                        &decrypted_size);
+
+    if (err != THEMIS_SUCCESS) {
+        fprintf(stderr, "failed to decrypt message: %d\n", err);
+        return 3;
+    }
+
+    /*
+     * And finally check that the result is the same as the input.
+     */
+
+    if (decrypted_size != message_size) {
+        fprintf(stderr,
+                "message length does not match: actual %zu, expected %zu\n",
+                decrypted_size,
+                message_size);
+        abort();
+    }
+    if (memcmp(message_bytes, decrypted_bytes, message_size) != 0) {
```

Comment:
haha, this line (and similar lines below) will trigger all SAST tools again :D
(which is totally fine, just funny)

Reply:
> haha, this line (and similar lines below) will trigger all SAST tools again :D

But why? `memcpy()` is perfectly fine for comparing bytes...

Well, maybe they *could be* triggered to suggest a constant-time comparison here, but I don't want to link the entire OpenSSL here just for that, which will not yield any new issues being noticed by AFL. For one, this binary is leaking memory and it's fine (ASAN detects it, but AFL tells it to shut up).

Reply:
> they could be triggered to suggest a constant-time comparison here

exactly, because SAST tools are usually quite primitive and they don't care if memory comparison is happen on sensitive data or not.